packages/protocol/contracts/L2/DelegateOwner.sol

```diff
@@ -72,63 +83,49 @@ contract DelegateOwner is EssentialContract, IMessageInvocable {
         // This contract's owner will be itself.
         __Essential_init(address(this), _addressManager);
 
-        if (_realOwner == address(0) || _l1ChainId == 0 || _l1ChainId == block.chainid) {
+        if (_remoteOwner == address(0) || _remoteChainId == 0 || _remoteChainId == block.chainid) {
             revert DO_INVALID_PARAM();
         }
 
-        l1ChainId = _l1ChainId;
-        realOwner = _realOwner;
+        remoteChainId = _remoteChainId;
+        remoteOwner = _remoteOwner;
         admin = _admin;
     }
 
     /// @inheritdoc IMessageInvocable
-    function onMessageInvocation(bytes calldata _data)
-        external
-        payable
-        onlyFromNamed(LibStrings.B_BRIDGE)
-    {
-        IBridge.Context memory ctx = IBridge(msg.sender).context();
-        if (ctx.srcChainId != l1ChainId || ctx.from != realOwner) {
-            revert DO_PERMISSION_DENIED();
-        }
+    function onMessageInvocation(bytes calldata _data) external payable onlyAuthorized {
         _invokeCall(_data, true);
     }
 
-    /// @dev Invokes a call by the admin
-    /// @param _data The data for this contract to interpret.
-    function invokeCall(bytes calldata _data) external {
-        if (msg.sender != admin) revert DO_PERMISSION_DENIED();
-        _invokeCall(_data, true);
+    /// @notice Dryruns a message invocation but always revert.
+    /// If this tx is reverted with DO_TRY_RUN_SUCCEEDED, the try run is successful.
+    /// Note that this function shall not be used in transaction and is designed for offchain
+    /// simulation only.
+    function dryrunInvocation(bytes calldata _data) external payable {
+        _invokeCall(_data, false);
+        revert DO_DRYRUN_SUCCEEDED();
     }
 
     /// @dev Updates the admin address.
     /// @param _admin The new admin address.
-    function setAdmin(address _admin) external {
-        if (msg.sender != owner() && msg.sender != admin) revert DO_PERMISSION_DENIED();
-        if (admin == _admin) revert DO_INVALID_PARAM();
+    function setAdmin(address _admin) external nonReentrant onlyFromSelf {
```

Comment:
I think onlyFromSelf only true in case it is a `delegateCall` via `multicall`, right ?
But what if the "remoteOwner" or admin wants to set this ? Maybe we dont want "removeOwner" to set this, but i guess admin can do it indeed, without the need of a multicall, but "simple bridge call" ?

Reply:
- you can see delegatecall as borrowing code from another contract into this (DelegateOwner) contract. 
- calling `setAdmin` in the multicall3 (as a delegate call) will make `msg.sender` to be `address(this)`.

The change in this PR is to make sure the only entrance to call anything is `onMessageInvocation`, and enforce `setAdmoin` and `acceptOwnership` to be called as inner transations.

Reply:
>     * calling `setAdmin` in the multicall3 (as a delegate call) will make `msg.sender` to be `address(this)`.
> The change in this PR is to make sure the only entrance to call anything is `onMessageInvocation`, and enforce `setAdmin` and `acceptOwnership` to be called as inner transations.

Yes, this is what i'm questioning/asking. Do we want to allow to **only** be called via bridge+multicall ? So no direct call ever, neither 'simple' bridge transaction (without multicall) ?

Reply:
Direct calling `onMessageInvocation` by the local admin is still allowed.